// @framer/twitter.twitter-ui/code/Profile_Image.tsx
import * as React from "react"
import { Frame, Size, PropertyControls, ControlType } from "framer"

interface Props {
    username: string
    parentSize: Size
    radius: number
}

interface State {
    username: string
    radius: number
}

export class ProfileImage extends React.Component<Partial<Props>, State> {
    static defaultProps = {
        username: "design",
        radius: "50%"
        width: 64,
        height: 64
    }

    static propertyControls: PropertyControls<Props> = {
        username: { type: ControlType.String, title: "Username" },

    }

    state = {
        username: ProfileImage.defaultProps.username,
    }

    componentDidMount() {
        const { username, radius } = this.props
        this.setState({ username, radius })
    }

    componentWillReceiveProps(props: Props) {
        if (props.username !== this.props.username) {
            this.setState({ username: props.username, radius: props.radius })
        }
    }

    render() {
        const { username, radius } = this.props
        var ProfileImageURL = `https://twitter.com/${username}/profile_image?size=original`;

        return (
            <img src={ProfileImageURL} style={{ ...style, borderRadius: radius }} />
        )
    }
}

const style: React.CSSProperties = {
    position: "absolute",
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    width: "100%",
    height: "100%"
}
