// @framer/twitter.twitter-ui/code/QR.tsx
import * as React from "react";
import { Frame, PropertyControls, ControlType } from "framer";
import QRCode from "react-qr-code";

// Define type of property
interface Props {
  bgColor: string;
  colorCount: number;
}

export class QR extends React.Component<Partial<Props>> {
  // Set default properties
  static defaultProps = {
    username: "design",
    bgColor: "#1B95E0",
    colorCount: 0,
  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    username: { type: ControlType.String, title: "Username" },
  };

  state = {
    bgColor: QR.defaultProps.bgColor,
    colorCount: QR.defaultProps.colorCount,
  };

  QRColors = ["#1B95E0", "#FFAD1F", "#794BC4", "#17BF63", "#F45D22", "#E0245E"];

  onTap = () => {
    let currentColor = this.state.colorCount;
    this.setState({ bgColor: this.QRColors[currentColor] });
    if (currentColor < this.QRColors.length - 1) {
      currentColor++;
    } else {
      currentColor = 0;
    }
    this.setState({ colorCount: currentColor });
  };

  render() {
    return (
      <Frame
        style={{
          height: "100%",
          width: "100%",
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
          background: this.state.bgColor,
          overflow: "hidden",
          borderRadius: "5px",
          borderWidth: "4px",
          borderStyle: "solid",
          borderColor: "#FFFFFF",
          minWidth: "200px",
          minHeight: "200px",
        }}
        onTap={this.onTap}
      >
        <QRCode
          value={`https://twitter.com/${this.props.username}`}
          bgColor={this.state.bgColor}
          fgColor="#FFFFFF"
          size={152}
          level="Q"
        />
        <div
          style={{
            height: "50px",
            width: "50px",
            background: this.state.bgColor,
            position: "absolute",
          }}
        >
          <img
            src={`https://twitter.com/${
              this.props.username
            }/profile_image?size=original`}
            style={{
              height: "80%",
              width: "80%",
              borderRadius: "50%",
              position: "absolute",
              top: "10%",
              left: "10%",
            }}
          />
          <div
            style={{
              width: "32px",
              height: "32px",
              display: "flex",
              alignItems: "center",
              justifyContent: "center",
              position: "absolute",
              top: "145%",
              left: "148%",
              background: this.state.bgColor,
            }}
          >
            <svg
              version="1.1"
              width="24"
              height="24"
              viewBox="0 0 250 204"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                fill="white"
                d="m78.6 203.6c94.3 0 145.9-78.2 145.9-145.9 0-2.2 0-4.4-0.1-6.6 10-7.2 18.7-16.3 25.6-26.6-9.2 4.1-19.1 6.8-29.5 8.1 10.6-6.3 18.7-16.4 22.6-28.4-9.9 5.9-20.9 10.1-32.6 12.4-9.4-10-22.7-16.2-37.4-16.2-28.3 0-51.3 23-51.3 51.3 0 4 0.5 7.9 1.3 11.7-42.6-2.1-80.4-22.6-105.7-53.6-4.4 7.6-6.9 16.4-6.9 25.8 0 17.8 9.1 33.5 22.8 42.7-8.4-0.3-16.3-2.6-23.2-6.4v0.7c0 24.8 17.7 45.6 41.1 50.3-4.3 1.2-8.8 1.8-13.5 1.8-3.3 0-6.5-0.3-9.6-0.9 6.5 20.4 25.5 35.2 47.9 35.6-17.6 13.8-39.7 22-63.7 22-4.1 0-8.2-0.2-12.2-0.7 22.6 14.4 49.6 22.9 78.5 22.9"
              />
            </svg>
          </div>
        </div>
      </Frame>
    );
  }
}
