// @framer/twitter.twitter-ui/code/Sprite.tsx
import * as React from "react";
import { Frame, PropertyControls, ControlType } from "framer";

interface Props {
  animated: boolean,
  url: string,
  numFrames: number,
}

interface State {
  animated: boolean
}

export class Sprite extends React.Component<Partial<Props>, State> {

  static defaultProps = {
    width: 200,
    height: 200,
    url: "https://abs.twimg.com/a/1534259568/img/animations/web_heart_animation_edge.png",
    numFrames: 29
  }

  static propertyControls = {
    url: {
      type: ControlType.String,
      title: "Sprite URL",
    },
    numFrames: {
      type: ControlType.Number,
      title: "# of Frames",
    },
  }

  constructor(props) {
    super(props);
    this.state = { animated: false };
  }

  handleClick() {
    if (this.state.animated) {
      this.setState({ animated: false });
    } else {
      this.setState({ animated: true });
    }
  }


  render() {
    var img = new Image();
    img.addEventListener("load", function(){
      // Loading image
    });
    img.src = this.props.url;
    console.log(img.width + ' ' + img.height)


    let style = {
      width: `${(img.width/this.props.numFrames - 1)/2}px`,
      height: `${(img.height)/2}px`,
      backgroundImage: `url(${this.props.url})`,
      backgroundPosition: this.state.animated ? "100%" : "0%",
      backgroundRepeat: "no-repeat",
      backgroundSize: `${this.props.numFrames * 100}%`,
      transition: this.state.animated ? `background-position 1s steps(${this.props.numFrames - 1})` : "",
      backgroundColor: 'transparent',
    }

    return (
      <div style={{
        height: "100%",
        width: "100%",
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
      }}>
        <Frame onTap={() => this.handleClick()} style={style} />
      </div>
      
    )
  }
}
