// @framer/ismael.google-maps/code/GoogleMap.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";

const style: React.CSSProperties = {
    height: "100%",
    width: "100%",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    overflow: "hidden",
    border: 0
};

interface Props {
    lat: number;
    lon: number;
    zoom: number;
}

export class GoogleMap extends React.Component<Props> {
    static defaultProps = {
        lat: -34.903667,
        lon: -56.1701647,
        zoom: 15
    }

    static propertyControls: PropertyControls = {
        lat: { type: ControlType.Number, step: 0.000001, min: -90, max: 90,  title: "Latitude" },
        lon: { type: ControlType.Number, step: 0.000001, min: -180, max: 180, title: "Longitude" },
        zoom: { type: ControlType.Number, step: 1, min: 0, max: 25, title: "Zoom" },
    }

    render() {
        const { lat, lon, zoom } = this.props;

        return (
            <iframe style={style} src={`http://maps.google.com/maps?q=${lat},${lon}&z=${zoom}&output=embed`}></iframe>
        );
    }
}
